### ISSUE TITLE
network mode "custom_network" not supported by buildkit

### ISSUE THREAD

**Issue by MarcosMorelli** [docker/buildx#175]
Background:
Running a simple integration test fails with **network** option:

 `docker network create custom_network`
`docker run -d --network custom_network --name mongo mongo:3.6`
`docker buildx build --network custom_network --target=test .`

Output:
`network mode "custom_network" not supported by buildkit`

Still not supported? Code related:
https://github.com/docker/buildx/blob/master/build/build.go#L462-L463

**Comment by tonistiigi** [comment]
Sorry, I'm not sure if we will ever start supporting this as it makes the build dependant on the configuration of a specific node and limits the build to a single node.

**Comment by bryanhuntesl** [comment]
> Sorry, I'm not sure if we will ever start supporting this as it makes the build dependant on the configuration of a specific node and limits the build to a single node.

That horse has bolted - SSH mount makes the build dependent upon the configuration of a single node - where did that dogma even get started?

**Comment by tonistiigi** [comment]
> That horse has bolted - SSH mount makes the build dependent upon the configuration of a single node

No, it does not. You can forward your ssh agent against any node or a cluster of nodes in buildx. Not really different than just using private images.

**Comment by bryanhuntesl** [comment]
> > That horse has bolted - SSH mount makes the build dependent upon the configuration of a single node
> 
> No, it does not. You can forward your ssh agent against any node or a cluster of nodes in buildx. Not really different than just using private images.

Why would someone do that? ssh-agent is a something that needs to be fairly well locked down - why would someone forward it across an insecure connection?

I mean, that's a tangent anyway. Being able to run integration-tests in a docker build was an incredibly useful feature, one less VM to spin up, and one less iceberg to melt, it's just useful because it's efficient. 

It's also great to not have to run nodejs, ruby, etc on the build host but instead just have them as container dependency, if you can do all your tests in a docker build container it's one less thing to lock down. 

Anyhow, I apologise for running off on a tangent. All I'm saying is, it would be awesome if you could bring that functionality into the latest version of docker along with the means to temporary mount secrets. It's just a really lightweight way to run disposable VMs without touching the host or even giving any rights to run any scripts or anything on the host.

**Comment by tonistiigi** [comment]
> why would someone forward it across an insecure connection?

Why would that connection be insecure? Forwarding agent is more secure than build secrets because your nodes never get access to your keys.

> if you can do all your tests in a docker build container it's one less thing to lock down.
>  along with the means to temporary mount secrets

We have solutions for build secrets, privileged execution modes (where you needed `docker run` before for more complicated integration tests) and persistent cache for your apt/npm cache etc. https://github.com/moby/buildkit/issues/1337 is implementing sidecar containers support. None of this breaks the portability of the build. And if you really want it, host networking is available for you.

**Comment by bryanhuntesl** [comment]
> None of this breaks the portability of the build. And if you really want it, host networking is available for you.

But I'd like to spin up a network for each build - and have all the stuff running that would be needed for the integration tests. But again, I have to loop back around and either do weird stuff with iptables, or run postgres on the host and share it with all builds (contention/secrets/writing to the same resources/etc). 

You could see how it would be so much more encapsulated and attractive if I could spin up a network per build with a bunch of stub services and tear it down afterwards ?

**Comment by bryanhuntesl** [comment]
> > why would someone forward it across an insecure connection?
> 
> Why would that connection be insecure? Forwarding agent is more secure than build secrets because your nodes never get access to your keys.

I'm talking about the socat hack where you forward the socket over TCP - you might have been referring  to something else.

**Comment by bryanhuntesl** [comment]
https://github.com/moby/buildkit/issues/1337 sounds cool but honestly, given the choice between something that right now works or something that will drop in 2 years time, I know what most of the community would choose.

**Comment by tonistiigi** [comment]
> you might have been referring to something else.

https://medium.com/@tonistiigi/build-secrets-and-ssh-forwarding-in-docker-18-09-ae8161d066

**Comment by bryanhuntesl** [comment]
> > you might have been referring to something else.
> 
> https://medium.com/@tonistiigi/build-secrets-and-ssh-forwarding-in-docker-18-09-ae8161d066

Nah your secrets and forwarding feature is great - love it. Rocker had secrets support 3 years ago but that project withered on the vine.

**Comment by bryanhuntesl** [comment]
The sidecar also sounds great and very clever and well structured. But again, 3 years ago I could build with secrets and talk to network services to run integration tests.

**Comment by alanreyesv** [comment]
> The sidecar also sounds great and very clever and well structured. But again, 3 years ago I could build with secrets and talk to network services to run integration tests.

Also, it does work in compose while build secrets does not.

**Comment by sudo-bmitch** [comment]
Adding another use case where specifying the network would be useful: "hermetic builds".

I'm defining a docker network with `--internal` that has one other container on the network, a proxy that is providing all the external libraries and files needed for the build. I'd like the docker build to run on this network without access to the external internet, but with access to that proxy.

I can do this with the classic docker build today, or I can create an entire VM with the appropriate network settings, perhaps it would also work if I setup a DinD instance, but it would be useful for buildkit to support this natively.

**Comment by bryanhuntesl** [comment]
> Adding another use case where specifying the network would be useful: "hermetic builds".
> 
> I'm defining a docker network with `--internal` that has one other container on the network, a proxy that is providing all the external libraries and files needed for the build. I'd like the docker build to run on this network without access to the external internet, but with access to that proxy.
> 
> I can do this with the classic docker build today, or I can create an entire VM with the appropriate network settings, perhaps it would also work if I setup a DinD instance, but it would be useful for buildkit to support this natively.

Good point, I should have mentioned I was doing that too for [git dependencies](https://github.com/kurzdigital/git-cache-http-server-docker), and... Docker themselves [have blogged about using it to augment the docker cache](https://docs.docker.com/samples/apt-cacher-ng). Now I just burn the network, take lots of coffee breaks, and do my bit to melt the ice caps.

**Comment by eriksw** [comment]
This is particularly needed in environments such as Google Cloud Build where ambient credentials (via special-IP metadata service) are available only on a particular named network, _not_ on the default network, in order to keep their exposure to build steps opt-in.

**Comment by existere** [comment]
> The recommendation is to use `buildx create --driver-opt network=custom` instead when you absolutely need this capability. The same applies to the google cloud build use case.

Anyone have a working example of this in Github Actions? Not working for me.

```sh
Run docker/setup-buildx-action@v1
  with:
    install: true
    buildkitd-flags: --debug
    driver-opts: network=custom-network
    driver: docker-container
    use: true
  env:
    DOCKER_CLI_EXPERIMENTAL: enabled
Docker info
Creating a new builder instance
  /usr/bin/docker buildx create --name builder-3eaacab9-d53e-490c-9020-xxx --driver docker-container --driver-opt network=custom-network --buildkitd-flags --debug --use
  builder-3eaacab9-d53e-490c-9020-bae1d022b444
Booting builder
Setting buildx as default builder
Inspect builder
BuildKit version
  moby/buildkit:buildx-stable-1 => buildkitd github.com/moby/buildkit v0.9.3 8d2625494a6a3d413e3d875a2ff7xxx
```

```sh
Build
/usr/bin/docker build -f Dockerfile -t my_app:latest --network custom-network --target production .
time="2022-01-19T17:00:XYZ" level=warning msg="No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load"
error: network mode "custom-network" not supported by buildkit. You can define a custom network for your builder using the network driver-opt in buildx create.
Error: The process '/usr/bin/docker' failed with exit code 1
```

**Comment by crazy-max** [comment]
@existere https://github.com/docker/buildx/blob/master/docs/reference/buildx_create.md#use-a-custom-network

**Comment by existere** [comment]
> @existere https://github.com/docker/buildx/blob/master/docs/reference/buildx_create.md#use-a-custom-network

I don't see how that setup is any different that my configuration. Am I missing something?

> Use a custom network
$ docker network create foonet
$ docker buildx create --name builder --driver docker-container --driver-opt network=foonet --use
$ docker buildx inspect --bootstrap
$ docker inspect buildx_buildkit_builder0 --format={{.NetworkSettings.Networks}}
map[foonet:0xc00018c0c0]

```sh
/usr/bin/docker buildx create --name builder-3eaacab9-d53e-490c-9020-xxx --driver docker-container --driver-opt network=custom-network --buildkitd-flags --debug --use
```

Here's the network create:
```sh
/usr/bin/docker network create custom-network
35bb341a1786f50af6b7baf7853ffc46926b62739736e93709e320xxx
/usr/bin/docker run --name my_container --network custom-network 
```

**Comment by tonistiigi** [comment]
> I don't see how that setup is any different that my configuration

You don't pass the custom network name with build commands. Your builder instance is already part of that network.

**Comment by philomory** [comment]
OK, so once you've got it set up, how do you get name resolution to work? If I have a container `foo` that's running on my custom network, and I do `docker run --rm --network custom alpine ping -c 1 foo`, it's able to resolve the name `foo`. Likewise, if I create a builder with `docker buildx create --driver docker-container --driver-opt network=custom --name example --bootstrap`, and then `docker exec buildx_buildkit_example0 ping -c 1 foo`, that works. But if I have a Dockerfile with `RUN ping -c 1 foo` and then run `docker buildx build --builder example .`, I get `bad address foo`. If I manually specify the IP address, _that_ works, but hard-coding an IP address into the Dockerfile hardly seems reasonable.

**Comment by xkobal** [comment]
I have the same problem as @philomory. Name resolution doesn't work.
I am using `network=cloudbuild` on Gcloud platform, so I can't hardcode any IP address.

**Comment by fibbers** [comment]
It seems GCE's metadata server IP is `169.254.169.254` (but I'm not sure if this is always the case), so this worked for me in Google Cloud Build:
```sh
docker buildx create --name builder --driver docker-container --driver-opt network=cloudbuild --use
docker buildx build \
  --add-host metadata.google.internal:169.254.169.254 \
  --push \
  -t ... \
  .
```

and inside `Dockerfile`:
```dockerfile
RUN curl "http://metadata.google.internal/computeMetadata/v1/project/project-id" -H "Metadata-Flavor: Google"
```

**Comment by xkobal** [comment]
Thanks for the tips @fibbers, it works like a charm. It will do the job until a real fix.

**Comment by poconnor-lab49** [comment]
@tonistiigi What's the right way to use the `docker run` scenario you describe?

> We have solutions for build secrets, privileged execution modes (where you needed `docker run` before for more complicated integration tests) and persistent cache for your apt/npm cache etc. [moby/buildkit#1337](https://github.com/moby/buildkit/issues/1337) is implementing sidecar containers support. None of this breaks the portability of the build. And if you really want it, host networking is available for you.

I'm currently doing something like
```
# create network and container the build relies on
docker network create echo-server
docker run -d --name echo-server --network echo-server -p 8080:80 ealen/echo-server
# sanity check that the echo server is on the network
docker run --rm --network echo-server curlimages/curl http://echo-server:80

# create the Dockerfile, will need to hit echo-server during the build
cat << EOF > echo-client.docker
FROM curlimages/curl
RUN curl echo-server:80 && echo
EOF

# create the builder using the network from earlier
docker buildx create --name builder-5fa507d2-a5c6-4fb8-8a18-7340b233672e \
    --driver docker-container \
    --driver-opt network=echo-server \
    --buildkitd-flags '--allow-insecure-entitlement security.insecure --allow-insecure-entitlement network.host' \
    --use

# run the build, output to docker to sanity check
docker buildx build --file echo-client.docker \
    --add-host echo-server:$(docker inspect echo-server | jq '.[0].NetworkSettings.Networks["echo-server"].IPAddress' | tr -d '"\n') \
    --tag local/echo-test-buildx \
    --output type=docker \
    --builder builder-5fa507d2-a5c6-4fb8-8a18-7340b233672e .
```
Using `add-host` like this seems like a dirty hack just to reach another container on the same network. What would be the right way to do this?

**Comment by sudo-bmitch** [comment]
I've been seeing similar. You can run the build in a user specified network. But the buildkit container on that network has DNS set to the docker's localhost entry which won't get passed through to nested containers. So the RUN steps within the build don't have that DNS resolution. I'm not sure of the best way to get that to pass through, perhaps a proxy running in the buildkit container that lets DNS get set to the container IP instead of localhost?

**Comment by rsinnet** [comment]
If you're trying to access the GCE VM Metadata Server for authentication, it is possible. Apologies for not sharing this earlier. You're welcome to use this builder:

Docker Hub: https://github.com/MisoRobotics/cloudbuildx
Source: https://hub.docker.com/r/misorobotics/cloudbuildx

```yaml
name: misorobotics/cloudbuildx
args: [--tag=myimage, .]
```

Or `misorobotics/cloudbuildx:multiarch` if you want multiarch.

For the curious, this file has the meat of it: https://github.com/MisoRobotics/cloudbuildx/blob/main/entrypoint.sh

**Comment by dm17** [comment]
> The recommendation is to use `buildx create --driver-opt network=custom` instead when you absolutely need this capability. The same applies to the google cloud build use case.

How does this work in the case of docker-compose? All of my build time networking is done cleanly in docker-compose.ymls. Just found this thread; so far I've posted my context in the other thread about this: https://github.com/moby/buildkit/issues/978

**Comment by dm17** [comment]
I don't understand why we can't have custom networks during build time anymore. It also seems buildx, docker, and compose are increasingly incongruous. For example, now that I'm trying to use the buildx work around, I see the project, compose_project_name, and args from compose and dockerfile do not work with buildx (maybe there's another work around that I've not found yet, of course).

**Comment by TBBle** [comment]
> Why does it seem like bake is also incompatible with this? After creating the builder with the `--driver-opt` option and `--use`'ing it, `buildx bake` still says `ERROR: network mode "custom" not supported by buildkit`.

This might be a config mistake, or a bug in bake. `buildx bake` should be being told to use the builder you created on the custom network (if not default), but not itself being told about the custom network with `--network`.

If you can get your build working with `docker buildx build` but not `docker buildx bake` then I'd suggest opening a new issue with repro steps so any issue in bake can be tracked down and fixed.

**Comment by dm17** [comment]
> This might be a config mistake, or a bug in bake. `buildx bake` should be being told to use the builder you created on the custom network (if not default), but not itself being told about the custom network with `--network`.

Right, that's why I opened this a couple weeks ago: https://github.com/docker/buildx/issues/1761

> If you can get your build working with `docker buildx build` but not `docker buildx bake` then I'd suggest opening a new issue with repro steps so any issue in bake can be tracked down and fixed.

I haven't been able to actually but I need to put more effort into it. It just feels like too many nested workarounds at this point to depend on. What stopped me from `docker buildx build` was it couldn't find the first layer of images that were already created using another Dockerfile, which the Dockerfile `buildx build` starts `FROM`. It would just try to download them from `docker.io` and it also wouldn't read the `${project}` env var that has always worked in the Dockerfile.

**Comment by dm17** [comment]
I will send $100 to anyone who adds custom networks back to docker compose by default (without disabling BuildKit). Sorry it is low, but I'm not valued in the billions of dollars, unlike the Docker corporation.

I can't cope going in circles anymore trying to mitigate functionality issues in buildx/bake to work with compose files. It just overcomplicates everything in an already complicated environment which was previously working well. Until then I'll continue to disable BuildKit like the folks above. Considering moving my infra over to podman or just baremetal.

**Comment by spurin** [comment]
@dm17 - recommend opening a new issue and referencing this one. Within it, provide a sample docker-compose.yaml file and working/non-working command line examples (even if it's using older versions of compose).

You're more likely to get someone looking into this if they can easily recreate the scenario you're encountering 👍

It caught my attention and I'm intrigued, however, this is already quite a busy thread already making it a bit more challenging to fully understand how you're seeing this.

**Comment by dm17** [comment]
> docker swarm has the 'overlay' network capability

I don't think swarm addings any capability to builds, which is the issue here.

@spurin "You're more likely to get someone looking into this if they can easily recreate the scenario you're encountering"

There's nothing to reproduce really. It should be clear to everyone that custom networks during build time used to work (due to them being supported in the classic builder) and now no longer work (due to BuildKit not supporting them).

**Comment by Hronom** [comment]
this is crazy, `docker compose` was used to spin up dependencies e.g. db, redis etc. But `build-push-action` not able to connect to that docker containers. This is critical...

I try use:
```
          driver-opts: |
            network=tmp_default
```
in `docker/setup-buildx-action@v3`

and then run `docker/metadata-action@v5` and during build it not able to resolve host name for `db` and `redis`.

What should I do?

**Comment by TBBle** [comment]
That's not nearly enough information to get help, I fear. Is this repository public? Then we could actually see what you've tried to do, and possibly work out what's gone wrong.

**Comment by Hronom** [comment]
In threads that has started 4 years ago and still not have valid solution I disable polite. 

I would suggest maintainers of this repo read this thread that started 4 years ago. 
I spend yesterday around 30 min - 1 hour and I was able to read all comments here and  in this thread and restore context.

Since this repo in docker org, my hope that [docker org](https://www.docker.com/company/) has developers/managers/directors/architects on salary and be able to spend 30 min - 1 hour to read comments to get context and understand the issue that still present in buildx that becomes default now(who validated functionality of it?).

This is very sad story to see such threads. 

There very common and popular use case that blocked by this: you build your app using multi stage build and you also wanna run tests that used DB, Redis, Elasticsearch etc, so you laucnh DB/Redis/Elasticsearch using docker compose and you wanna conenct to them from your tests that you run during the build.

It's doubles the sadness, because it was supported previously.

**Comment by Hronom** [comment]
Ok, so for those who is looking for some solution like me, here I created repo https://github.com/Hronom/buildx-add-host-example with workaround tha is aggregated from this thread and from [this topic](https://twl.lab49.com/author/poconnor-lab49/afaa80643a69ababdb1308fdb7b7ee9c)

@poconnor-lab49 big respect to you! Thanks for the inspiration in [this topic](https://twl.lab49.com/author/poconnor-lab49/afaa80643a69ababdb1308fdb7b7ee9c)

Shame to buildkit/buildx that you still not able to find normal solution within 4 years for common and popular use case. This is really sad story.

**Comment by dm17** [comment]
Indeed @Hronom, it feels like gaslighting when they act like 
1) why would you ever need to do this 
2) you can still do it via some work around that doesn't actually work (buildx) 
3) no one should need to do this (despite it being a feature of previously working software)
4) ignore this thread for so many years

**Comment by Zahlii** [comment]
I am currently hitting this issue, too, with the following setup on my Jenkins. I want to a) spin up a postgres docker image b) build a python library inside a Dockerfile, while running tests against said postgres database with a fixed name.


The issue is that my company wants me to use the docker plugin for Jenkins (https://plugins.jenkins.io/docker-workflow/, see https://docs.cloudbees.com/docs/cloudbees-ci/latest/pipelines/docker-workflow)


The Jenkinsfile code looks similar to this here:

    docker network create flow-network
    docker.image('postgres:15.4-bookworm').withRun('--network=flow-network --name postgres-db...') {
        c ->
        docker.build(TAG, "--network flow-network .") // This will run python code assuming DB available at postgres-db
      }


Now, I could rewrite this code to work with buildx as said above, but then I'd need to use basic shell syntax as opposed to the plugin, which will perform clean-up activities in case of failures automatically.

**Comment by jorismak** [comment]
My pipelines are still working fine for some time, but trying to create a new container locally I now notice that `docker build` is not working anymore. Apparently `docker build` is an alias for `buildx` (Whatever that is) but buildx can't have a simple 'network' flag to build with a network? (The recommendation is still to disable the default network on local installs and create a custom bridge network, so this seems quite essential??).

So, to get a build going locally, I now have to 'create a buildx node' with the network flag, then tell buildx to use that node, then when use 'buildx build' instead of just 'build', and the first thing it does is load up some buildx-image.

.. why? If the build system is being replaced by buildx, at least make it seamless and feature-parity before doing things like this (or make users opt in to the experimental stuff).

Making the `docker build` commandline compatible by including an option to set the network to use makes it already a lot better, but I'm sure other people have other issues with it reading this thread.

**Comment by 64J0** [comment]
I was able to reach the containers using `--network "host"` but this is not good enough, since some people that would run this command are not using Linux, and this flag does not work in other major OS' like Mac and Windows (even with WSL).

**Comment by 64J0** [comment]
So, this other approach apparently worked for me:

```bash
# start the necessary container using docker-compose
# so it already creates the network and attaches the container to it
# this container exposes a port and is attached to the network some-network
docker-compose up -d some-database

# get the gateway IP for the some-network
SOME_NETWORK_GW_IP=$(docker container inspect some-database --format='{{ $network := index .NetworkSettings.Networks "some-network" }}{{ $network.Gateway }}')

docker build . --add-host "some-database:${SOME_NETWORK_GW_IP}" --tag "some-api" --target "some-api" --file "Dockerfile.api"
```

**Comment by jorismak** [comment]
I think it's a lot of different workarounds , for something that could be
simple .

With the buildx plug-in (which is in the install docs these days ) the
'build' command doesn't manage to pass the '--network' parameter along to
the buildx builder being automatically created.

And since the '--network' parameter exists and tries to do _something_ with
buildx it seems like a bug to me, but that's probably the view from my
limited bubble.

**Comment by 64J0** [comment]
Indeed, ideally it would be more simple. These workarounds are ugly as hell.

**Comment by VladimirAndrianov96** [comment]
I found a nice workaround, it's also relevant to any other frontend framework too, short workaround strategy is this:
1. There is a `host` network mode available, meaning by default the build happens in `sandboxed` isolated mode, so we switch our Next.js app build time network to host mode from the isolation
2. we're in the `host` network, so follow this rule: ```From your container's perspective localhost is the container itself. If you want to communicate with the host machine, you will need the IP address of the host itself. Depending on your Operating System and Docker configuration, this IP address varies. host.docker.internal is recommended from Docker version 18.03 onward.``` 
3. Use `host.docker.internal` as Strapi app hostname, if we take into account Strapi+Postgres container up and running with `:1337` port exposed for Strapi we need connection string value like this
`NEXT_PUBLIC_STRAPI_API_URL=http://host.docker.internal:1337/graphql`

That way we have build and runtime both working just fine, without external network setup, just two independent docker-compose files started locally on the same machine
It works like that:

* Next.js build network driver is switched to `host` mode from the `sandboxed` mode -> it builds and asks graphql for data -> connects to `host.docker.internal:1337` -> Docker's internal network redirects it to `localhost:1337` -> Docker routes this to locally exposed Strapi running at `:1337` port -> connection is fine

so, frontend build stage can connect straight to internet  to fetch cloud CMS instance or `host.docker.internal` connection to access another local docker-compose setup,  but it can't connect directly to `localhost:1337` or attach to custom network and use DNS there like `http://strapi:1337/`, so the build stage network setup is very limited and needs some tweaking like this, only `https://cms-in-the-cloud.com/graphql` or only `http://host.docker.internal:1337/graphql`

![image](https://github.com/docker/buildx/assets/15718242/a2d1b10d-b0d5-43cf-b091-4a960faea40a)
There are only two ways to achieve the desired result with new Buildkit engine used for Docker
Option one: Use CMS deployment from the cloud and connect Next.js during build phase to it
Option two: Use local CMS deployment from different network, but connect to it via host.docker.internal

This will work for you on the Windows machine with latest Docker version used (Docker v3 and above with Buildkit engine), for Linux/MacOs consider to use 127.0.0.1 address instead if you have older versions of Docker

To make this work follow the external network configuration at Docker like this

Docker-compose A for Next.js
***Specify .env connection to Strapi like this ```NEXT_PUBLIC_STRAPI_API_URL=http://host.docker.internal:1337/graphql```***

```
version: '3.8'
services:
  webnextjs:
    container_name: webnextjs
    ....
    ....
    build:
      network: host
    ports:
      - '3000:3000'
```
      
Docker-compose B for Strapi + PostgreSQL for data storage
```
version: "3.8"

services:
  strapiDB:
    container_name: strapiDB
      ....
      ....
      ports:
        - "5432:5432"
      expose:
        - "5432"
      networks:
        - strapiPostgresNetwork

  strapi:
    container_name: strapi
    ....
    ....
    ports:
      - "1337:1337"
    expose:
      - "1337"
    depends_on:
      strapiDB:
    networks:
      strapiPostgresNetwork:

networks:
  strapiPostgresNetwork:
    driver: bridge
    name: strapiPostgresNetwork
```      


Run the Docker-compose B to get Strapi up, and then run the Docker-compose A to run the Next.js build+run process that will go through the `build` and `run` phases of it nicely through the 
***```NEXT_PUBLIC_STRAPI_API_URL=http://host.docker.internal:1337/graphql```*** connection, that's it!

More docs:

Read https://github.com/docker/buildx/issues/591#issuecomment-816843505 to see ```Buildkit supports none, host, default(sandboxed) as network mode. Supporting custom network names is currently not planned as it conflicts with the portability and security guarantees.``` and don't use any other custom networks with Buildkit, that's the new reality of ```docker build``` command. 

In short, custom network cannot be specified for the build step of the docker-compose anymore, meaning that connecting two docker-compose files into a single ***custom*** external network to have build running as before is not feasible anymore after the Docker moved to this https://docs.docker.com/build/buildkit/ engine which was enabled by-default in v23.0.0 on 2023-02-01 (https://docs.docker.com/engine/release-notes/23.0/#2300)

With custom network **runtime phase** is always fine but this Docker v3 update breaks the **build phase** completely making the build impossible if **any `getStaticProps` static asset** can't be fetched during the **build phase** ```(docker build)``` and fails the build so it **never** reaches the **runtime phase** ```(docker run)``` (if you skip build and got to runtime, it will work)
* although the Docker's official compose file guide still lists this option as this was added in version 3.4+ format specification https://docs.docker.com/compose/compose-file/compose-file-v3/#network

![image](https://github.com/docker/buildx/assets/15718242/71bb858b-d2c6-466a-af06-ace2b9133e38)

The use case for this setup:
Your organization has `website/Strapi` and `website/Nextjs` repositories separately, but you just want to run them both locally in same Docker network running `website/Strapi` docker-compose file and `website/Nextjs` docker-compose file with **build step,** this is stupid simple use case we all use in everyday life and this is where Docker v3 introduced that breaking change.

Any other solution I found is to use `docker buildx ...` but that way you lose the `docker-compose` features or merge strapi and next.js in one big monorepo with big single docker-compose file

**Comment by dm17** [comment]
Thank you @VladimirAndrianov96 - it is a great summary of how they're wasting our time by removing features they claim we didn't want/need.

Does anyone know if podman/podman-compose has custom networks during build time? I'm ready to take the leap, despite the unpaid dev hours it'll take.

**Comment by jorismak** [comment]
> I found a nice workaround, it's also relevant to any other frontend framework too, short workaround strategy is this:

Any workaround which involves 'finding out the IP of' is not acceptable, since that absolutely does not make it portable :). That works maybe for a local setup where you are the only one using the docker files and you know _your_ host. In any kind of network setting or team setting, that is out of the window.

Also, can't be 'the way it is supposed to be'

> Read [#591 (comment)](https://github.com/docker/buildx/issues/591#issuecomment-816843505) to see `Buildkit supports none, host, default(sandboxed) as network mode. Supporting custom network names is currently not planned as it conflicts with the portability and security guarantees.` and don't use any other custom networks with Buildkit, that's the new reality of `docker build` command.

The thing is that they are still recommending (and it's kind of needed with how it work) to disable the builtin bridge network as step 1 after installing Docker Engine. So you need to create custom networks and supply them to any container or build command that starts and needs internet or some sort of container-to-container network mode. 
Breaking the `--network` command from `docker build` (since it's passed blindly to buildx which uses a different network parameter all together) means breaking the build command. 

I know a lot of people here are commenting about how it works with docker-compose, but it just basically breaks the default `docker build` command. How....

The workaround is to create your own builder, and apply a network to that. But I would much rather just see the 'buildx' enforcement being removed, because it seems completely useless to me if you are just building things locally to then push.
But that's also not a good workaround if you often switch networks for your build commands (as you do in a docker-compose file with custom networks).


> 
> Any other solution I found is to use `docker buildx ...` but that way you lose the `docker-compose` features or merge strapi and next.js in one big monorepo with big single docker-compose file

In theory, this means you use docker-compose to build and then launch? Simply building in a separate step would work I guess. I don't use docker-compose here. Development is done locally and I'm the one making the images for deployment for the team, so I only work local and push to our registry.



If someone can tell me howto go back to the normal old 'docker build' command and just disable and ignore the whole buildx thing, yes please.

**Comment by dm17** [comment]
If someone has a Hacker News account with street/interweb cred, then perhaps they can post this thread on there... It is often how such situations get attention from companies with loads of money whose sole purpose is supposedly to remedy such situations - rather than cause them.

**Comment by jedevc** [comment]
> If someone has a Hacker News account with street/interweb cred, then perhaps they can post this thread on 

Heya, so, I'm no longer at docker and so only really speak for myself here, but I do feel the need as a maintainer on this project to call this out.

This is an open source project. Contributions and contributors are welcome - any person in this thread is welcome to engage on *how* to fix this issue, and upstream any contributions. It's known to be a limitation, but demanding that the maintainers drop everything to work on this is just not great form, sorry. Just like you, the maintainers of this project are busy, having to deal with internal/external constraints on their time, and often other things take priority.

That said, if you're a paid user of docker and this is affecting you, and you need to see this resolved, then you should raise this through support/your account representative. They can escalate internally and help set engineering priority - this is generally true of any corporate-backed open source project.

Finally, essentially calling for a brigade of a github thread *during the break between Christmas and the new year, a time which very often people take off from work and would not be able to respond* is... not great. I sincerely hope no good open source citizen would do this, but if they had, it would have the potential to disrupt people's time with family and friends. Please don't do that.

Please let's move this thread back on track - sharing workarounds and discussing ways in which this could be worked on productively.

**Comment by dm17** [comment]
I'm not surprised that someone with pronouns in their GitHub profile just creates an entire victim narrative out of nothing. However, I am thankful for [jedevc](https://github.com/jedevc) for pointing out that I should try pushing on a docker account rep for this issue.

Calling it a "brigade" is *your* narrative - not a fact. Especially when loads of stuff has gotten attention via this method, there is no rule against posting in a more general developer publication (which HN is), and after the attention was gotten via that method no one retrospectively said "I know billion dollar corporation X helped us only after thread Y was posted, but we now realize this was actually a brigade."

I will also add "do not post anything that causes workers not in your own company or family to scramble unnecessarily during the holidays" to my post - despite the fact that I never said the opposite, but I suppose it could be implied by the fact that I made the suggestion on Dec 29th. I view this as an async rather than sync comm.

**Comment by dm17** [comment]
> Please let's move this thread back on track - sharing workarounds and discussing ways in which this could be worked on productively.

It was blocked immediately by this:

> Sorry, I'm not sure if we will ever start supporting this as it makes the build dependant on the configuration of a specific node and limits the build to a single node.

Where is the evidence for this claim? Also, it seems to me that this person is on Docker's payroll - and still I am not asking them to stop their holiday and work @jedevc - I'm merely replying with a reasonable question in a session of asynchronous communication. They can choose to use the technology they own so they only see work messages during work hours, for example. This above comment also implies that the fix for this issue will not be accepted by Docker, so it definitely dissuades anyone not on the docker payroll from attempting one using their precious time.

**Comment by thaJeztah** [comment]
> I'm not surprised that someone with pronouns in their GitHub profile just created an entire victim narrative out of this. 

Please keep such comments at the door. No need for this.

I'll skip over some of you other wordings, because I don't think there's anything constructive in those

> Where is the evidence for this claim?

There's many things associated with this, and this feature is not trivial to implement (if possible at all). Let me try to do a short summary from the top of my head. Note that I'm not a BuildKit maintainer; I'm familiar with the overal aspects, but not _deeply_ familiar with all parts of BuildKit.


First of all, buildx (the repository this ticket is posted in) is a _client_ for BuildKit; the feature requested here would have to be implemented by the builder (BuildKit); Ideally this ticket would be moved to a more appropriate issue tracker, but unfortunately GitHub does not allow transferring tickets between orgs; in addition, _multiple_ projects may need to be involved. There's no harm in having a ticket in this repository (nor to have discussions elsewhere), but ultimately this feature would have to be both accepted by upstream BuildKit and/or Moby maintainers, and implemented respectivee projects.

Now the more technical aspect of this feature request (or: this is where the fun starts).

BuildKit can be run in different settings / ways; 

- _standalone_, (bare-metal, containerized, in a kubernetes cluster); Buildx provides commands to instance new builders (`docker buildx create`) which can create new containerized builders
- as part of the Moby daemon ("Docker Engine") as the embeded / built-in builder

Depending on the above, semantics, and available features will differ.

### Standalone BuildKit builders

Standalone builders are designed to be stateless; builders may have cache from prior builds, but don't persist images (only build-cache), nor do they have a concept of "networks" (other than "host" networking or "no" networking). The general concept here is that standalone builders can be ephemeral (auto-scaling cluster), and that builds don't depend on prior state. They _may_ advertise certain capabilities (BuildKit API version, native platform/architecture), but otherwise builders are interchangeable (perform a build, and export the result (e.g. push to a registry, or export as an OCI bundle).

Given that standalone builders don't have access to "docker custom networks" (there's no docker daemon involved), it won't be possible to provide a _per build_ option to use a specific network. The _workaround_ mentioned in this thread is to;

- create a _containerized_ builder (BuildKit daemon runing in a container)
- to run that container with specific docker network attached.
- to run builds _inside_ the container with "host" networking; i.e., run in the host's networking namespace where in this setup "host" means the container in which the BuildKit daemon runs.

In this configuration, the "host" (container) can resolve containers running in that custom network, but this _only_ works in very specific scenarios;

- the builder-container is running on a Docker Engine (can't be supported on bare-metal BuildKit daemons, nor on Kubernetes or other drivers)
- the Docker Engine must already have the given custom network (standalone buildkit has no means to control the Docker Engine to create the network)
- dependencies must be manages through other means; other containers that must be resolvable must be attached on the custom network, and must be running before thee build is started
- _all builds_ on the builder must use the same custom network; using a different custom network means the builder must be either destroyed/recreated (or detached and re-attached to another network); which also means it cannot be used for parallel builds depending on different networks.
- if builds affect state in dependency containers (e.g. depend on a database-container), all parts of a build must run on the same node (and potentially must be serialized, not in parallel) 

All of the above combined make this a _workaround_ that would work in only very specific scenarios. Implementing this as a native feature for standalone builders would quickly go down the rabbit-hole; "custom network attached to this builder" as well as "state of dependencies" would have to be exposed as capabilities, so that buildx can query which builder to select for a specific build, or an external orchestrator would need to be involved to construct builders (and dependencies) on-demand. Both would be moving away significantly from the current design of standalone builders (stateless, ephemeral).

### As part of the Moby daemon ("Docker Engine")

When using the "default' builder on Docker Engine, BuildKit is not running as a standalone daemon, but compiled into the Docker Engine. In this scenario, BuildKit has some (but limited) access to features provided by the Docker Engine. For example, it's possible to use images that are available in the local image cache as part of the build (e.g. an image you built earlier, but that's not pushed to a registry). There's also limitations; when using "graphdrivers", the Docker Engine does not provide a multi-arch image store, so it's not possible to build multple architectures in a single build. (This will be possible in the near future and  being worked on as part of the [containerd image store integration](https://docs.docker.com/storage/containerd/)).

Containers created during build are optimized for performance; containers used for build-steps tend to be very short-lived. "regular" containers as created by the Docker Engine have a non-insignificant overhead to provide all features that `docker run` can provide. To reduce this overhead, BuildKit creates optimized containers with a subset of those features; to further improve performance, it also skips some intermediate processes: BuildKit acts as its own runtime, and it can directly use the OCI runtime (runc) without requiring the intermediate "docker engine", and "containerd" processes.

Long story short; build-time-containers are (by design) isolated from "regular" containers; they're not managed by the docker daemon itself, won't show up in `docker ps`, and networking is not managed by the regular networking stack (no DNS-entries are registered in the internal DNS). 

So, while the "embedded" BuildKit may have more _potential_ options to integrate with custom networks, integrating with custom networks will require a _significant_ amount of work; both in BuildKit (integration with the network stack) and in Moby / "docker engine" to (somehow) allow BuildKit to create different "flavors" of containers (isolated / non-isolated). This will come with a performance penalty, in addition to complexity involved (a secondary set of (short-lived) containers that can be attached to a network, but are not directly managed by the Moby / Docker Engine itself.

**Comment by dm17** [comment]
Thank you for the analysis.
Do you think the classic builder will be deprecated in the near future within docker, docker compose or buildx?

**Comment by TafkaMax** [comment]
TLDR; Which scenario would be better?

**Comment by TBBle** [comment]
Thank you @thaJeztah for that clarifying summary.

I wonder if it'd be worth a feature request against Docker Compose, which seems like it can do most of what the "containerised builder workaround" needs, since people in this thread have tried to do it that way already.

If Docker Compose was able to create/use/remove temporary private builder instances on its own (or the config-specified) network, along with resolving the apparent issue that a standalone BuildKit instance in a Docker container attached to a custom network doesn't get the right DNS setup for services on that custom network (see https://github.com/docker/buildx/issues/175#issuecomment-1732566254 from earlier attempts to use the "workaround" manually), then I think it can deliver the "workaround" flow fairly naturally for some of the use-cases described here.

I see a similar idea mentioned at https://github.com/docker/compose/pull/10745#issuecomment-1609671858 but I think that was about selecting an existing builder, rather than instantiating a new one. https://github.com/compose-spec/compose-spec/issues/386 is along these lines but it wants to be super-generic; it is probably too generic for this use-case, particularly if we want to tear-down the builder instance after usage or at compose-down time. In this case, the builder is more like another service in Docker Compose that compose knows how to use when building service images. (That also might be a better way to visualise and implement it, similar to existing `depends_on` semantics, and semantics for defining service deployment etc. in the Compose file.)

That is separate from the existing `--builder` flag for `docker compose build` introduced in Docker Compose 2.20.0 in July 2023.

That said, I'm not a Docker Compose user, so I may be overestimating what it can cover, or misunderstanding the relevant workflow. Either way, unless this is an obviously-faulty idea to Docker Compose users, it'd be better to discuss in a ticket there, to focus on the relevant use-cases that involve Docker Compose, and also focus this ticket on buildx-involved use-cases.

**Comment by thaJeztah** [comment]
> Do you think the classic builder will be deprecated in the near future within docker, docker compose or buildx?

There's no _immediate_ plans to actively remove the classic builder, but no active development is happening on it; consider it in "maintenance mode", and mostly to support building native Windows containers (BuildKit does not _yet_ support Windows Containers, although work on that is in progress). The classic Builder's architecture does not give a lot of room for further expansion, so it will start to diverge / get behind BuildKit more over time. The classic builder may also not make the transition to the [containerd image-store integration](https://docs.docker.com/storage/containerd/); there's currently a very rudimentary implementation to help the transition, but we're aware of various limitations in that implementations that may not be addressable with the classic buidler.

> EDIT: For me personally - build time is not the primary thing I am after. I wish the solution to work. Currently using the classic builder still works, but my primary concern is that this feature disappears.

Perhaps it'd be useful to start a GitHub discussion; GitHub tickets aren't "great" for longer conversations and don't provide threads (not sure which repository would be best; perhaps BuildKit (https://github.com/moby/buildkit/discussions) to collect slightly more in-depth information about use-cases. I know there was a lot of contention around the original implementation, which also brought up concerns about portability and the feature being out of scope for _building_ (see lengthy discussion on https://github.com/moby/moby/issues/10324 and https://github.com/moby/moby/pull/20987 (carried in https://github.com/moby/moby/pull/27702)).

Use-cases that I'm aware of;

- (ab)use `docker build` as a CI pipeline; spin up a stack of containers to test/verify things in an intermediate stage before tagging the final stage
- caching; e.g. run a apt-mirror / [Apt-Cacher](https://help.ubuntu.com/community/Apt-Cacher%20NG) in a container to locally host packages used for the build

But there may be other use-cases. Some of those may make more sense in a "controlled" environment (local / special purpose machine; single user), but get complicated fast in other environments. Having more data about use-cases could potentially help design around those (which _may_ be through a different approach).

**Comment by sudo-bmitch** [comment]
To repeat a nearly 2 year old comment here, buildkit does support running with a custom network, it's even documented: https://docs.docker.com/build/drivers/docker-container/#custom-network

The issue isn't that it won't run on a custom network, instead, as so often happens on the internet, it was DNS. When you run a container runtime (which buildkit does) inside of a container on a custom network, it sees the DNS settings of that parent container:

```shell
$ docker network create build
9b7c83ceda7e2552e99d27c29d275936e882fd9cc9488361209bbf4421c2f180

$ docker run -it --rm --net build busybox cat /etc/resolv.conf
search lan
nameserver 127.0.0.11
options ndots:0
```

And as docker and other runtimes do, they refuse to use 127.0.0.11 as a DNS server, so the nested container falls back to 8.8.8.8. Here's a demo for proof:

```shell
#!/bin/sh

set -ex

docker network create custom-network

echo "hello from custom network" >test.txt
docker run --name "test-server" --net custom-network -d --rm \
  -v "$(pwd)/test.txt:/usr/share/nginx/html/test.txt:ro" nginx
server_ip="$(docker container inspect test-server --format "{{ (index .NetworkSettings.Networks \"custom-network\").IPAddress }}")"
echo "Server IP is ${server_ip}"

cat >Dockerfile <<EOF
FROM curlimages/curl as build
USER root
ARG server_ip
RUN mkdir -p /output \
 && cp /etc/resolv.conf /output/resolv.conf \
 && echo \${server_ip} >/output/server_ip.txt \
 && (curl -sSL http://test-server/test.txt >/output/by-dns.txt 2>&1 || :) \
 && (curl -sSL http://\${server_ip}/test.txt >/output/by-ip.txt 2>&1 || :)

FROM scratch
COPY --from=build /output /
EOF

docker buildx create \
  --name custom-net-build \
  --driver docker-container \
  --driver-opt "network=custom-network"
docker buildx build --builder custom-net-build --build-arg "server_ip=${server_ip}" \
  -o "type=local,dest=output" .

docker buildx rm custom-net-build
docker stop test-server
docker network rm custom-network
```

Running that shows that custom networks are supported, just not DNS:

```shell
$ ./demo-custom-network.sh
+ docker network create custom-network
bd5ce7361f5fc94b0da0fe32a3f5482176a6fcaca68997556d3449269c451cea
+ echo hello from custom network
+ pwd
+ docker run --name test-server --net custom-network -d --rm -v /home/bmitch/data/docker/buildkit-network/test.txt:/usr/share/nginx/html/test.txt:ro nginx
31bf9516fdc6dba7d97796be6b6c55f2a134a3050a7b1286a2ac96658e444c62
+ docker container inspect test-server --format {{ (index .NetworkSettings.Networks "custom-network").IPAddress }}
+ server_ip=192.168.74.2
+ echo Server IP is 192.168.74.2
Server IP is 192.168.74.2
+ cat
+ docker buildx create --name custom-net-build --driver docker-container --driver-opt network=custom-network
custom-net-build
+ docker buildx build --builder custom-net-build --build-arg server_ip=192.168.74.2 -o type=local,dest=output .
[+] Building 4.2s (9/9) FINISHED
 => [internal] booting buildkit                                                                    1.8s
 => => pulling image moby/buildkit:buildx-stable-1                                                 0.4s
 => => creating container buildx_buildkit_custom-net-build0                                        1.5s
 => [internal] load build definition from Dockerfile                                               0.0s
 => => transferring dockerfile: 393B                                                               0.0s
 => [internal] load metadata for docker.io/curlimages/curl:latest                                  0.6s
 => [auth] curlimages/curl:pull token for registry-1.docker.io                                     0.0s
 => [internal] load .dockerignore                                                                  0.0s
 => => transferring context: 2B                                                                    0.0s
 => [build 1/2] FROM docker.io/curlimages/curl:latest@sha256:4bfa3e2c0164fb103fb9bfd4dc956facce32  1.2s
 => => resolve docker.io/curlimages/curl:latest@sha256:4bfa3e2c0164fb103fb9bfd4dc956facce32b6c5d4  0.0s
 => => sha256:4ca545ee6d5db5c1170386eeb39b2ffe3bd46e5d4a73a9acbebc805f19607eb3 42B / 42B           0.1s
 => => sha256:fcad2432d35a50de75d71a26d674352950ae2f9de77cb34155bdb570f49b5fc3 4.04MB / 4.04MB     0.8s
 => => sha256:c926b61bad3b94ae7351bafd0c184c159ebf0643b085f7ef1d47ecdc7316833c 3.40MB / 3.40MB     0.8s
 => => extracting sha256:c926b61bad3b94ae7351bafd0c184c159ebf0643b085f7ef1d47ecdc7316833c          0.1s
 => => extracting sha256:fcad2432d35a50de75d71a26d674352950ae2f9de77cb34155bdb570f49b5fc3          0.1s
 => => extracting sha256:4ca545ee6d5db5c1170386eeb39b2ffe3bd46e5d4a73a9acbebc805f19607eb3          0.0s
 => [build 2/2] RUN mkdir -p /output  && cp /etc/resolv.conf /output/resolv.conf  && echo 192.168  0.2s
 => [stage-1 1/1] COPY --from=build /output /                                                      0.0s
 => exporting to client directory                                                                  0.0s
 => => copying files 357B                                                                          0.0s
+ docker buildx rm custom-net-build
custom-net-build removed
+ docker stop test-server
test-server
+ docker network rm custom-network
custom-network

$ cat output/resolv.conf
search lan
options ndots:0

nameserver 8.8.8.8
nameserver 8.8.4.4
nameserver 2001:4860:4860::8888
nameserver 2001:4860:4860::8844

$ cat output/server_ip.txt
192.168.74.2

$ cat output/by-dns.txt
curl: (6) Could not resolve host: test-server

$ cat output/by-ip.txt
hello from custom network
```

**Comment by thaJeztah** [comment]
> And as docker and other runtimes do, they refuse to use 127.0.0.11 as a DNS server, so the nested container falls back to 8.8.8.8. Here's a demo for proof:

Hmm.. right, but that should _not_ be the case when using `--network=host`. In that case, the container should inherit the `/etc/resolv.conf` from the host; here's a docker-in-docker running with a custom network (`127.0.0.11` is the ambedded DNS resolver); a container with `--network=host` inherits the host's settings;

```bash
cat /etc/resolv.conf
nameserver 127.0.0.11
options ndots:0

docker run --rm --network=host alpine cat /etc/resolv.conf
nameserver 127.0.0.11
options ndots:0
```

Trying to do the same with a BuildKit container builder that has `host` networking allowed, and a build started with `--network=host` shows that BuildKit does _not_ do the same; it uses the default DNS servers;

Creating a custom network and a "test-server" container attached to it;

```bash
docker network create custom-network
docker run -d --name test-server --network custom-network nginx:alpine
docker container inspect test-server --format '{{ (index .NetworkSettings.Networks "custom-network").IPAddress }}'
172.24.0.2
```

Create a custom builder attached to the network, and allow "host-mode" networking;

```bash
docker buildx create --name custom-net-build --driver docker-container --driver-opt network=custom-network --buildkitd-flags '--allow-insecure-entitlement network.host'
custom-net-build
```


Running a build with `--network=host`, which _should_ run in the host's networking namespace _and_ inherit the host's DNS configuration (`127.0.0.11` - the embedded DNS);

```bash
docker buildx build --no-cache --builder custom-net-build --network=host --progress=plain --load -<<'EOF'
FROM alpine
RUN cat /etc/resolv.conf
RUN wget http://test-server
EOF
```

However, it looks like BuildKit sets the default DNS resolvers, not inheriting from the host;

```
#4 [1/3] FROM docker.io/library/alpine:latest@sha256:51b67269f354137895d43f3b3d810bfacd3945438e94dc5ac55fdac340352f48
#4 resolve docker.io/library/alpine:latest@sha256:51b67269f354137895d43f3b3d810bfacd3945438e94dc5ac55fdac340352f48 done
#4 CACHED

#5 [2/3] RUN cat /etc/resolv.conf
#5 0.038 options ndots:0
#5 0.038
#5 0.038 nameserver 8.8.8.8
#5 0.038 nameserver 8.8.4.4
#5 0.038 nameserver 2001:4860:4860::8888
#5 0.038 nameserver 2001:4860:4860::8844
#5 DONE 0.0s

#6 [3/3] RUN wget http://test-server
#6 0.057 wget: bad address 'test-server'
#6 ERROR: process "/bin/sh -c wget http://test-server" did not complete successfully: exit code: 1
```


So I think there's something funky going on there, and BuildKit's executor / runtime does not take `host` networking into account for DNS resolvers 🤔

**Comment by TBBle** [comment]
BuildKit incorrectly replacing localhost DNS resolvers when using host networking is https://github.com/moby/buildkit/issues/3210. There was [a PR in progress](https://github.com/moby/buildkit/pull/3244) just over a year ago, but it wasn't completed. https://github.com/moby/buildkit/issues/2404 seems to have had more recent activity, but looks much wider in scope than https://github.com/moby/buildkit/issues/3210.

**Comment by crazy-max** [comment]
> BuildKit incorrectly replacing localhost DNS resolvers when using host networking is [moby/buildkit#3210](https://github.com/moby/buildkit/issues/3210). There was [a PR in progress](https://github.com/moby/buildkit/pull/3244) just over a year ago, but it wasn't completed. [moby/buildkit#2404](https://github.com/moby/buildkit/issues/2404) seems to have had more recent activity, but looks much wider in scope than [moby/buildkit#3210](https://github.com/moby/buildkit/issues/3210).

Should be solved with https://github.com/moby/buildkit/pull/4524

**Comment by TBBle** [comment]
If you want to test the fixed BuildKit _right now_, create a builder with `--driver-opt=image=moby/buildkit:master,network=custom-network <other params>`.

In fact, I made that change in the shell script from https://github.com/docker/buildx/issues/175#issuecomment-1875486526 on line 30

```sh
--driver-opt "image=moby/buildkit:master,network=custom-network" \
```

and it failed the same way; and I confirmed that it was running BuildKit from commit 2873353.

I can see that the BuildKit worker is supposed to be using host-mode networking:

```
Labels:
 org.mobyproject.buildkit.worker.executor:         oci
 org.mobyproject.buildkit.worker.hostname:         d1295746541b
 org.mobyproject.buildkit.worker.network:          host
 org.mobyproject.buildkit.worker.oci.process-mode: sandbox
 org.mobyproject.buildkit.worker.selinux.enabled:  false
 org.mobyproject.buildkit.worker.snapshotter:      overlayfs
```

So I tried with the extra changes from https://github.com/docker/buildx/issues/175#issuecomment-1875540193, resulting in
```sh
docker buildx create \
  --name custom-net-build \
  --driver docker-container \
  --driver-opt "image=moby/buildkit:master,network=custom-network" \
  --buildkitd-flags "--allow-insecure-entitlement network.host"
docker buildx build --builder custom-net-build --build-arg "server_ip=${server_ip}" \
  --network host \
  -o "type=local,dest=output" .
```

and that worked correctly:

```console
$ cat output/resolv.conf
nameserver 127.0.0.11
options ndots:0
$ cat output/by-dns.txt
hello from custom network
```

So it'd be nice if the buildx docker-container driver could automatically set up the hosted buildkit with `--network host`-equivalent, either when using a custom network, or when the network mode is not CNI (defaulted or explicitly "host").

Technically, I guess if BuildKit used the worker's network config rather than the specific `RUN` command's network config when making resolv.conf decisions, then it'd work too, since clearly the custom network is _reachable_ even without `--network host`.

I'm not 100% clear on the network layering here. Maybe the `RUN`'s container (without `--network=host`) is actually being attached to the custom network too, rather than inheriting from its parent, and so BuildKit is still doing the wrong thing by not producing a resolv.conf that matches this behaviour?

Anyway, the relevant buildkit change should be part of the 0.13 release and any pre-releases _after_ 0.13b3, and automatically picked up by `docker-container` drivers through the default `moby/buildkit:buildx-stable-1` tag, updated when [the BuildKit maintainers bless](https://github.com/moby/buildkit/actions/workflows/buildx-image.yml) a build as stable-enough.

**Comment by TBBle** [comment]
Okay, results of discussion with BuildKit maintainer on that PR is that making this smoother is a BuildX thing, as BuildKit is changing its default network config such that `host` mode will no longer be the implicit default.

For reference, here is my current _working_ mod of https://github.com/docker/buildx/issues/175#issuecomment-1875486526

<details>

```sh
#!/bin/sh

set -ex

docker network create custom-network

echo "hello from custom network" >test.txt
docker run --name "test-server" --net custom-network -d --rm \
  -v "$(pwd)/test.txt:/usr/share/nginx/html/test.txt:ro" nginx
server_ip="$(docker container inspect test-server --format "{{ (index .NetworkSettings.Networks \"custom-network\").IPAddress }}")"
echo "Server IP is ${server_ip}"

cat >Dockerfile <<EOF
FROM curlimages/curl as build
USER root
ARG server_ip
ADD http://${server_ip}/test.txt /output/by-ip-add.txt
ADD http://test-server/test.txt /output/by-dns-add.txt
RUN mkdir -p /output \
 && cp /etc/resolv.conf /output/resolv.conf \
 && echo \${server_ip} >/output/server_ip.txt \
 && (curl -sSL http://test-server/test.txt >/output/by-dns.txt 2>&1 || :) \
 && (curl -sSL http://\${server_ip}/test.txt >/output/by-ip.txt 2>&1 || :)

FROM scratch
COPY --from=build /output /
EOF

docker buildx create \
  --name custom-net-build \
  --driver docker-container \
  --driver-opt "image=moby/buildkit:master,network=custom-network" \
  --buildkitd-flags "--allow-insecure-entitlement=network.host --oci-worker-net=host"
docker buildx build --builder custom-net-build --build-arg "server_ip=${server_ip}" \
  --network host \
  -o "type=local,dest=output" .
docker buildx inspect custom-net-build

docker buildx rm custom-net-build
docker stop test-server
docker network rm custom-network

set +x
for d in output/by-*; do echo -n "$d:"; cat $d; done
```
</details>

This should remain working even when BuildKit changes default networking to bridge, as I've explicitly passed "--oci-worker-net=host" to the buildkitd in the container.

I also added a pair of ADD calls to the Dockerfile, to demonstrate that even if you remove `--network host` from the `docker buildx build` call, they can still the custom network DNS as they operate in buildkitd's network namespace, i.e. implicitly `host`. Once `bridge` becomes the default, that will break. (I tried to have a quick test of what happens when bridge-mode becomes the default, but [hit a minor issue with the buildkit image](https://github.com/moby/buildkit/issues/4637))

For buildx's docker-container driver, a simple nicety would be for the "custom network" driver option to automatically enable those two buildkitd flags `--allow-insecure-entitlement=network.host --oci-worker-net=host` since without the former, you can't pass `--network host` to `docker buildx build`, and without the latter, the default BuildKit network mode change will break builds that are working today by only using the custom network for `ADD` and `COPY`.

(Thinking about this, the kubernetes driver must be doing something similar, unless the same problem shows up there...)

Then the [buildx docker-container custom network docs](https://docs.docker.com/build/drivers/docker-container/#custom-network) also needs to mention that you need to use `docker buildx build --network host` to access your custom network from `RUN` commands. That's a little semantically weird but it is documentable. Nicety options there are welcome, but I note that the buildkit maintainer feedback was that simply automatically adding `--network host` was not a good choice.

Longer term/more-fully, buildx could perhaps use the new bridge mode to expose the custom network to both the buildkitd (`ADD`/`COPY`) and workers (`RUN`), removing the need to do anything special in the `docker buildx build` command except choose the already-configured-for-custom-network builder instance, and presumably mildly improving the network isolation in the process.

**Comment by tonistiigi** [comment]
https://github.com/docker/buildx/issues/2255
https://github.com/docker/buildx/issues/2256

**Comment by dm17** [comment]
@TBBle Does that mean that one would be able to use `bake` with a compose file?

**Comment by TBBle** [comment]
I _believe_ so, yes. AFAIK Bake is just driving these same components underneath, and I _think_ all the relevant configuration options as used in the test-case I extended are exposed in ComposeFile; but note that I am not a Compose user so I haven't tried this, and some of the ComposeFile fields will not be usable as they will be passing parameters to `docker buildx build` that instead need to go to `docker buildx create`.

**Comment by dm17** [comment]
@TBBle Thanks; that's also what I found. I appreciate the suggestions but don't yet have the mental energy earmarked in order to complete any. I'll wait a little longer in hopes that someone else streamlines a solution before attempting again :)

**Comment by MattiasMartens** [comment]
My two cents about use cases.

I was led here after trying to collapse a complex CI process into a Dockerfile. Specifically, a step runs migrations on a DB, another step dumps the DB's schema to a file, and a third step generates templates using that output. All this work follows the mandate of Docker build recipes—repeatable, portable—but the implementation of spinning up a DB doesn't fit with the paradigm.

Reason being: while migrations can run on the DB in step one, the DB itself must be running and accessible for step two, which has to run on a different image. By design, you can't keep a build layer alive once you've moved on to another one. Thus, in my use case, it points to a reasonable need for the DB to be attached as an external service, but still scoped and isolated to the build process.

Regrettably, the current tooling ecosystem _almost_ serves this purpose. It provides a `network` parameter, but won't allow custom networks. There is a workaround using buildx containers, but in my experimentation, that workaround did not work-around (code not available; I could not reach the container that should allegedly have been attached).

Seeking support for the issue, one then reaches this labyrinth of support tickets. And the first thing one sees there, confusingly, is the proverbial “slap of the wrist” for wanting or expecting the feature to exist in the first place.

As I see it: Docker builds are supposed to be able to network. We know this. To avoid unpredictable build outputs, the user should use the internet only to download resources that are themselves invariant for a given set of build parameters. If the user breaks this rule, they have earned the consequences.

Why should resources isolated to a custom network be any different? Docker networks are an extension of the general concept of networking, and in this context an irreducible one. One of the tolls you pay for moving off of bare metal is that containers need a mediating layer if they want to be able to talk to each other in an un-ambiguous way. It's natural to expect that tool to be available wherever it is needed, and build-time is one of the places where it is sometimes needed.

What I’m saying is: this isn’t really about repeatability or portability, is it? It’s about difficulty of implementation. I appreciate that it takes a vast amount of work to satisfy a particular use-case without breaking something else, but that makes it all the more unfortunate to see effort be wasted on framing a difficult use-case as somehow un-“Dockerly”. Especially because it is only difficult as a result of competing priorities taking precedence (BuildKit cache optimizations etc) and especially because of CLI, docs, and GitHub leading developers down a garden path that sometimes ends at that _extremely unsatisfying_ answer.

**Comment by TBBle** [comment]
It's not merely difficulty of implementation; if that were the case, someone would have done it by now. _I_ probably would have done it by now if it was merely "hard to implement".

The driving change is that with the introduction of BuildKit and buildx, the structure of the container build system has shifted to be much more strongly isolated, both from outside influences and particularly from Docker.

_The build system is no longer Docker-specific._ This has been part of Docker's direction for years now, as seen by the break-out of moby/moby, OCI standardisation of container formats, etc.

The upshot of that for our purposes is that _Docker's internal overlay network implementation_ is not visible to BuildKit (`docker buildx build`), unless you specifically create a builder instance on that custom network (`docker buildx create --driver docker-container --driver-opt "network=my-custom-network"`), in the same way if you want to give your builds access to some Kubernetes environment, you create a builder instance using the Kubernetes driver. Even the buildkit instance compiled into Docker doesn't have full access to the repertoire of Docker features, as described in the last segment of https://github.com/docker/buildx/issues/175#issuecomment-1874749665.

So wanting to have your builds run against a custom network is not being rejected as "Possible but unDockerly". However, to do that in the current state, you actually _need_ to do it, Docker is no longer in a position to make that work magically for you.

So "using buildx containers" is not a work-around, that's _the way_ to achieve what you want. As it's the way to achieve anything you want from buildx that isn't available via the current built-in, limited, (and default) buildkit integration.

There's definitely space to improve the UX on "using buildx containers", but it's still going to be the same rough flow of "To do something that we can't do from the built-in buildkit builder, create a custom builder that supports what you need, use it, and tear it down when done" with varying levels of automation around it. Docker and `buildx build` directly probably can't do much about this, but orchestration tools like buildx bake or docker compose probably have scope to wrap this sort of thing up in their configuration files. (Although right now, neither actually offers to manage services used during build _at all_, so that's a larger discussion to be had).

**Comment by dm17** [comment]
@MattiasMartens Very well said. I would just add that it was so easy and simple previously. You don't know you miss something until it is deprecated! I must say I don't follow @TBBle's thought process as well as Mattias's. First, because I don't consider it magic to make a network available to a container. And second because "the structure of the container build system has shifted to be much more strongly isolated" means isolated from what? Clearly not the internet. I'm not claiming I'm owed a more streamlined compose experience, but I do feel it is important to correctly communicate the experience - as Mattias has. This design short circuits the UX of docker compose for users that run containers that needed access to other containers at build time. What happens, for example, when docker detects that a newer version of the container should be running from `compose up` and attempts to build it? Then the build will fail because `compose up` cannot exercise the requisite code path(s), right?

**Comment by MattiasMartens** [comment]
@TBBle If I understand correctly, you’re saying it’s not merely “hard to implement” but _impossible_ to implement because of a decision about separation of concerns made by BuildKit?

I think I follow your explanation broadly but there are some unfortunate choices of phrase... For example "Docker is no longer in a position to make [custom networks in the build environment] work magically for you" is a strange way to phrase "Docker no longer allows one feature, provided by Docker, through the CLI, to leverage another feature, provided by Docker, through the CLI". If that’s "magic" then we are all wizards!

I would be much happier with the status quo if someone, in the five years of this issue being open, had reported success in binding their custom network to their build environment by leveraging a custom builder. So far everyone who has tried this, including myself, has reported failure. Consider this an open request for a working example.

**Comment by sudo-bmitch** [comment]
@MattiasMartens 
>  Specifically, a step runs migrations on a DB, another step dumps the DB's schema to a file, and a third step generates templates using that output. All this work follows the mandate of Docker build recipes—repeatable, portable—but the implementation of spinning up a DB doesn't fit with the paradigm.
> 
> Reason being: while migrations can run on the DB in step one, the DB itself must be running and accessible for step two, which has to run on a different image. By design, you can't keep a build layer alive once you've moved on to another one. Thus, in my use case, it points to a reasonable need for the DB to be attached as an external service, but still scoped and isolated to the build process.

This application design is strongly at odds with the general best practices of microservices design and the reason that there is so much friction attempting to implement it within Docker. It implies that the state of the packaged application also includes state in an external and independent database. That tight linkage of state creates issues when upgrading and downgrading container images independent of the database.

Specifically for this issue, I'd propose that the DB should not be needed to convert migration steps into a database schema file, and that tooling that could convert between the two would significantly improve the build process for that application. If that's feasibly impossible, then starting a database, running the migration, and dumping the schema, could be done in a single step that runs the database on localhost in the background, eliminating the need for a network.

I get that this is counter to the design of some popular application frameworks. The difficulty is they are directly opposed to other efforts in the software development ecosystem, like the 12 Factor App, and now the desire to provide hermetic and reproducible builds.

> I would be much happier with the status quo if someone, in the five years of this issue being open, had reported success in binding their custom network to their build environment by leveraging a custom builder. So far everyone who has tried this, including myself, has reported failure. Consider this an open request for a working example.

I posted exactly that back in January in this very issue. What was missing from my example?

**Comment by MattiasMartens** [comment]
> I posted exactly that back in January in this very issue. What was missing from my example?

From a distance, your comment from January of this year (https://github.com/docker/buildx/issues/175#issuecomment-1875486526) could be read as a demonstration of the problem, not of a solution.

On a closer read, it does explain my issue. If my priorities were such that the ability to use custom networks in the build process were worth the cruft of putting `docker container inspect test-server --format "{{ (index .NetworkSettings.Networks \"custom-network\").IPAddress }}"` or other such incantation in my script, I would do it.

Maybe if this pattern gets wrapped into a Docker Compose plugin, that plugin could also do something like add the container's name and IP to `/etc/hosts` and take most of the pain out of the process.

> It implies that the state of the packaged application also includes state in an external and independent database.

In this case the starting point is a DB image. The only reason I can't apply migrations within the build script is because the DB image and the migration script image (which expects to be able to connect to a DB over the network) are two separate images, and to coordinate between them in the same dockerfile would require inter-process communication between build stages. Hence instantiating the DB on a custom network as an attempted workaround.

However, the migration script image is really just a wrapper around a binary, so I’m realizing my best bet is to copy that binary into my DB image and run it there which cuts out the need for an external network. Good riddance!

**Comment by TBBle** [comment]
@dm17 

> And second because "the structure of the container build system has shifted to be much more strongly isolated" - meaning isolated from what exactly?

Per rest of my quoted sentence, "both from outside influences and particularly from Docker.".

Specifically, the build engine is no longer part of the Docker project, in the same way the container engine stopped being part of Docker and became containerd. This allows it to be useful to and contributed-to by a wider variety of projects and be able to be used in a wider variety of environments.

Speculatively, Docker saw the writing on the wall that things like Kubernetes were not going to keep Docker underlying them forever, so anything that _required_ Docker (like the class Docker builder) was going to fall by the wayside. You can see lots of attempts to build non-Docker container build projects, but having tried a fair few back in the day for scaling/distributed builds on k8s, buildkit has been my personal best-result, and you can even drive k8s-based buildkit instances from the Docker CLI etc using buildx, so it's also the least-painful way for users familiar with the Docker ecosystem to access that wider build ecosystem.

----

@MattiasMartens

> If I understand correctly, you’re saying it’s not merely “hard to implement” but _impossible_ to implement because of a decision about separation of concerns made by BuildKit?

Roughly, yes. I guess _maybe_ it'd be possible if someone managed to write a CNI plugin that could hook up an arbitrary container to the Docker libnetwork-driven overlay networks on the various platforms, and then work out how to bundle that inside the buildx or dockerd binary (which CNI probably makes impossible, as it relies on command-line interfaces and special binary name/paths to function).

However, it's possible (I haven't looked closely in a while) that the in-Docker buildkit instance can't use CNI with its OCI work for some other structural reason, even if such a CNI plugin was available.

Honestly, the _simplest_ code-solution to make `docker buildx build --network custom-network` work would probably be to recognise when that is used with the default builder, and instead spin up a `docker-container` builder with the appropriate driver-opt, use it, and then throw it away. It'd be slower that the built-in builder and probably hit a bunch of other niggles, but more-accurately mimic the way things used to work. Same for any other `docker build` command-line arguments that now require using a `docker-container` or other external builder. (I suspect something very like this is the likely way forward for ComposeFile support too, but am not certain as I'm mostly speculating on the desired compose workflow, I haven't really used Docker Compose heavily, and even then never with a custom network for container _builds_.)

Just to be clear, the decision wasn't made by BuildKit. BuildKit is one of the _results_ of those decisions.

> "[...] work magically for you"

I meant this in the sense that it used to be a single command-line or Compose-file option, and now you have to do approximately three commands to make it work. As seen by the years of comments on this and related tickets, that's clearly a significant UX difference. In my opinion, anything you can bundle up into a config file that drives a whole ecosystem of tools is "magic" compared to having to maintain shell scripts, and I don't think it's strange to say so.

> I would be much happier with the status quo if someone, in the five years of this issue being open, had reported success in binding their custom network to their build environment by leveraging a custom builder. So far everyone who has tried this, including myself, has reported failure. Consider this an open request for a working example.

I've posted _two_ working examples in this ticket just this year: One with [`docker buildx build` run from a shell script](https://github.com/docker/buildx/issues/175#issuecomment-1938036177) and one with [Docker compose and `docker buildx bake`](https://github.com/docker/buildx/issues/175#issuecomment-1941063965), adapted from other people's posted working examples.

And a quick check shows approximately 0 follow-up comments reporting failure to use those approaches.

**Comment by MattiasMartens** [comment]
Thanks @TBBle for the added context, I feel quite clear about this now.

The inability to leverage DNS for custom networks in the build environment is the final hurdle that might have tripped folks up. I’ve seen some failure reports _somewhere_, maybe not in this issue but somewhere in the 2 o'clock haze from last night. Presumably they didn’t try your example because that does make the solution quite explicit.

> Honestly, the simplest code-solution to make docker buildx build --network custom-network work would probably be to recognise when that is used with the default builder, and instead spin up a docker-container builder with the appropriate driver-opt, use it, and then throw it away. It'd be slower that the built-in builder and probably hit a bunch of other niggles, but more-accurately mimic the way things used to work.

I am in favour of this as a solution, particularly if it can somehow solve the DNS issue. I realize the solution I posted about above (https://github.com/docker/buildx/issues/175#issuecomment-2379790250)—modifying the /etc/hosts file for the build container—would have serious problems (not robust to container’s operating system, invalidates build cache layers each time a container or network is rebuilt)

**Comment by TBBle** [comment]
Which DNS issues do you mean?

There _was_ a long-standing bug with DNS access to the linked services due to BuildKit's resolv.conf mangling needing an opt-out for host-networking mode, but [that was fixed in February](https://github.com/docker/buildx/issues/175#issuecomment-1937491543) and was shipped in BuildKit 0.13. That's why I was using `image=moby/buildkit:master` in my examples, to test that fix and show that hostname-based lookups "just worked" now, compared to the demonstration of the problem in January, and removing the need for the `--add-host echo-server:$(docker inspect echo-server | jq '.[0].NetworkSettings.Networks["echo-server"].IPAddress' | tr -d '"\n')` workaround posted in  2022.

**Comment by MattiasMartens** [comment]
@TBBle I was referring to the hostname-based lookup issue. I see now that is fixed. Thank you.

The command I tried was `docker buildx create --name mybuilder --driver-opt network=my-network-name` so I was missing `--driver docker-container` and that’s presumably why it didn’t work, and/or the Buildkit image in use was pre-February.

**Comment by costa** [comment]
My usecase is a minimal personal container-based dev/test/stage environment where 'docker<b>—</b>compose' is the only "orchestration" (component-architecture) tool I need there; my git server is local to that "self server" environment, so, its component network is neatly included in many builds ([more info on the setup](https://medium.com/@costa/git-decentralized-a25f00fd2955#92a6)) -- and this issue here is one of the major reasons I'm not using 'docker compose' with buildkit this far. Thank you in advance for addressing it properly.

**Comment by gr3m1in** [comment]
Also require this option due to build it on restricted environment, despite what environment it will be "run" on.

**Comment by spandan-sharma** [comment]
I brought localstack into the mix thinking I'll unit test services in various compose stack without needing to mock either the db or the AWS layers and only if the tests pass, build the final image. IOW the tests run as part of the image building process in a multistage manner where the artifacts from the test stages aren't included in the production target but the production target depends on the test target passing. This eliminates the need for external orchestration and just prevents the build of a bad image altogether.

Only after setting everything up I realised I can't have the build process (the `RUN` which runs the tests) contact other local services on the same custom network :( . So I'm back to mocking calls to databases (or using sqlite etc) and AWS layers which are very tedious, (quite) error-prone, and not as elegant.

Edit: Not something useful to the thread, but just thought about adding that I completely understand @dm17 's frustration and felt the same anguish as they probably have (from their posts here).

**Comment by TafkaMax** [comment]
Coming back to this issue. I am trying to convert my CI to use the DOCKER_BUILDKIT=1. I have done stuff according to this step, but I am still getting the error.

I am also setting `docker context use`.

```bash
docker context create remote_dev_server --docker "host=ssh://$DEV_SERVER_USER@$DEV_SERVER_IP"
docker context use remote_dev_server
docker network inspect $NETWORK_NAME
docker buildx create --name dev_buildkit_builder --driver docker-container --driver-opt network=${NETWORK_NAME}
docker buildx use dev_buildkit_builder
docker buildx bake \
      --builder dev_buildkit_builder \
      --progress plain \
      --push
#0 building with "dev_buildkit_builder" instance using docker-container driver
#1 [internal] load local bake definitions
#1 reading docker-compose.yml 764B / 764B done
#1 DONE 0.0s
ERROR: network mode "REDACTED" not supported by buildkit - you can define a custom network for your builder using the network driver-opt in buildx create
```

This is running in `.gitlab-ci.yml`

What am I doing wrong?
